### PR TITLE
Make lookup-joins work with sub-queries 

### PR DESCRIPTION
**Pull request by mkleen**
## Summary of the changes / Why this improves CrateDB

This makes lookup-joins work with arbitrary sub-queries. It includes the following changes:

 -  Introduce `PlannedRelation` as place placeholder for `SelectSymbols` for logical plans
 -  Optimize sub-queries of `MultiPhase` with a optimizer pass in the `MultiPhaseExecutor`
 -  Relax the condition for applying this rule, when there is a difference of 5000 rows use the lookup-join query to make it work for more use-cases including https://github.com/crate/crate-issues/issues/2

```
Q: select count(*) from (select t1.id, t1.a, t1.b, max(t.c) from doc.t1 join (select id, c from doc.t2 where c > 1) t on t1.id = t.id group by t1.id, t1.a, t1.b) x;
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       39.319 ±   24.020 |     31.670 |     33.111 |     33.981 |    243.243 |
|   V2    |        5.719 ±   19.495 |      2.400 |      3.234 |      4.041 |    198.019 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               - 149.21%                           - 164.41%
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 149.21%
The test has statistical significance
```

See also https://github.com/crate/crate-benchmarks/pull/313

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/io/crate/planner/optimizer/rule/EquiJoinToLookupJoin.java [hunk 1]
```diff
@@ -83,38 +83,42 @@ public LogicalPlan apply(JoinPlan plan,
...
+            largerSide = rhs;
+        } else {
+            smallerSide = rhs;
+            largerSide = lhs;
+        }
+
+        Collect largerSideCollect;
+        if (largerSide instanceof Collect collect) {
+            largerSideCollect = collect;
+            smallerSide = planStats.memo().extract(smallerSide);
```
Comment (mkleen) [inline server/src/main/java/io/crate/planner/optimizer/rule/EquiJoinToLookupJoin.java:109]: The sub-query needs to be extracted and cannot contain group references because the memo cannot resolve it by the end of optimization cycle. The group-reference handling works over replace-sources.

**Comment by mkleen** [comment]
I think i prefer the version where the sub-queries become sources because it fits better with the design of the memo and the iterative optimizer. This is too much of an exception for my taste.

File: server/src/main/java/io/crate/planner/operators/LogicalPlanner.java [hunk 2]
```diff
@@ -647,6 +665,34 @@ public static void executeNodeOpTree(DependencyCarrier dependencies,
             .execute(consumer, txnCtx);
     }
 
+    private class MultiPhaseOptimizer extends LogicalPlanVisitor<PlannerContext, LogicalPlan> {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/operators/LogicalPlanner.java:668]: I had more in mind to do it here: 

https://github.com/crate/crate/blob/e668e759588f869a66d4c710d4a6c631b51ec78b/server/src/main/java/io/crate/execution/MultiPhaseExecutor.java#L51

That would have the advantage that the plan/sources doesn't need to be copied

File: server/src/main/java/io/crate/planner/optimizer/rule/EquiJoinToLookupJoin.java (hunk 1 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/planner/optimizer/rule/EquiJoinToLookupJoin.java:109]: Isn't `resolvePlan.apply` taking care of this? I don't quite follow

**Comment by mfussenegger** [comment]
> I think i prefer the version where the sub-queries become sources because it fits better with the design of the memo and the iterative optimizer. This is too much of an exception for my taste.

Adding them as sources is imho violating the semantics of the existing `sources` property.

File: server/src/main/java/io/crate/planner/optimizer/rule/EquiJoinToLookupJoin.java (hunk 1 above)
Comment (mkleen) [reply server/src/main/java/io/crate/planner/optimizer/rule/EquiJoinToLookupJoin.java:109]: The subquery needs to be completely recursively resolved, so that there is not group reference left. `resolvePlan.apply` does not do that. It only resolved the current group reference to the referenced logical plan.

Comment (mkleen) [reply server/src/main/java/io/crate/planner/optimizer/rule/EquiJoinToLookupJoin.java:109]: It could done with `resolvePlan.apply` but this requires that it is called recursively on the sub-query until all sub-queries are resolved.

**Comment by mkleen** [comment]
> Adding them as sources is imho violating the semantics of the existing `sources` property.

Ok, so we move with this approach forward ?

**Comment by mfussenegger** [comment]
> So we do not make the sub-queries of the Multiphase become sources and instead dereference the sub-queries in the EquiJoinToLookupJoin Rule from the Memo 

yes

> and optimize the sub-queries with a second pass?

Could we remove the first pass, and only do the optimization once late, right before the plan gets used?

**Comment by mkleen** [comment]
> Could we remove the first pass, and only do the optimization once late, right before the plan gets used?

ok!

File: server/src/main/java/io/crate/analyze/relations/PlannedRelation.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    @Override
+    public @NotNull List<Symbol> outputs() {
+        return source.outputs();
+    }
+
+    @Override
+    public String toString() {
+        return source.relationNames().getFirst().toString();
```
Comment (mkleen) [inline server/src/main/java/io/crate/analyze/relations/PlannedRelation.java:72]: What would be a good printable representation here? Show the full plan of the source ?

**Comment by mkleen** [comment]
retest this please

```
[io.crate.integrationtests.LogicalReplicationITest.test_subscription_state_order](https://jenkins.crate.io/job/CrateDB/job/crate_on_pr/14057/testReport/junit/io.crate.integrationtests/LogicalReplicationITest/test_subscription_state_order/)
```

File: server/src/main/java/io/crate/analyze/relations/PlannedRelation.java (hunk 3 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/analyze/relations/PlannedRelation.java:72]: I think this is fine as is, but could use `relationName().toString()`

File: server/src/main/java/io/crate/execution/MultiPhaseExecutor.java [hunk 4]
```diff
@@ -44,11 +45,13 @@ private MultiPhaseExecutor() {
     public static CompletableFuture<SubQueryResults> execute(Map<LogicalPlan, SelectSymbol> dependencies,
                                                              DependencyCarrier executor,
                                                              PlannerContext plannerContext,
-                                                             Row params) {
+                                                             Row params,
+                                                             BiFunction<LogicalPlan, PlannerContext, LogicalPlan> optimizer) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/MultiPhaseExecutor.java:49]:
```suggestion
                                                             BiFunction<LogicalPlan, PlannerContext, LogicalPlan> optimize) {
```

It's a function to apply, not a entity.

File: server/src/main/java/io/crate/planner/MultiPhasePlan.java [hunk 5]
```diff
@@ -75,7 +75,7 @@ public void executeOrFail(DependencyCarrier dependencyCarrier,
                               RowConsumer consumer,
                               Row params,
                               SubQueryResults subQueryResults) {
-        MultiPhaseExecutor.execute(dependencies, dependencyCarrier, plannerContext, params)
+        MultiPhaseExecutor.execute(dependencies, dependencyCarrier, plannerContext, params, (a,b) -> a)
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/MultiPhasePlan.java:78]: Shouldn't this pass along a optimize function?

File: server/src/main/java/io/crate/planner/operators/LogicalPlan.java [hunk 6]
```diff
@@ -213,7 +213,7 @@ default void executeOrFail(DependencyCarrier executor,
                                RowConsumer consumer,
                                Row params,
                                SubQueryResults subQueryResults) throws Exception {
-        LogicalPlanner.execute(this, executor, plannerContext, consumer, params, subQueryResults, false);
+        LogicalPlanner.execute(this, executor, plannerContext, consumer, params, subQueryResults, false, (a,b) -> a);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/operators/LogicalPlan.java:216]: Same here

File: server/src/main/java/io/crate/planner/operators/LogicalPlanner.java [hunk 7]
```diff
@@ -545,11 +560,13 @@ public static void execute(LogicalPlan logicalPlan,
                                RowConsumer consumer,
                                Row params,
                                SubQueryResults subQueryResults,
-                               boolean enableProfiling) {
+                               boolean enableProfiling,
+                               BiFunction<LogicalPlan, PlannerContext, LogicalPlan> optimizer
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/operators/LogicalPlanner.java:564]:
```suggestion
                               BiFunction<LogicalPlan, PlannerContext, LogicalPlan> optimize
```

File: server/src/main/java/io/crate/planner/optimizer/rule/EquiJoinToLookupJoin.java [hunk 8]
```diff
@@ -179,12 +177,12 @@ private static LogicalPlan createLookup(Collect smallerRelation,
...
         );
-        WhereClause whereClause = largerRelation.where().add(anyEqFunction);
-        LogicalPlan largerSideWithLookup = new Collect(largerRelation.relation(), largerRelation.outputs(), whereClause);
-        LogicalPlan smallerSideIdLookup = new Collect(smallerRelation.relation(), List.of(smallerRelationColumn), smallerRelation.where());
-        smallerSideIdLookup = new RootRelationBoundary(smallerSideIdLookup);
-
-        return MultiPhase.createIfNeeded(Map.of(smallerSideIdLookup, lookUpQuery), largerSideWithLookup);
+        WhereClause whereClause = largerSide.where().add(anyEqFunction);
+        LogicalPlan largerSideWithLookup = new Collect(largerSide.relation(), largerSide.outputs(), whereClause);
+        var smallerSideIdLookup = new RootRelationBoundary(smallerSide);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/optimizer/rule/EquiJoinToLookupJoin.java:182]: Couldn't we also already remove the `Collect` restriction for the larger side by wrapping it in a `Filter` operator instead of extending the `where()` ?

File: server/src/main/java/io/crate/planner/optimizer/rule/EquiJoinToLookupJoin.java [hunk 9]
```diff
@@ -179,12 +177,12 @@ private static LogicalPlan createLookup(Collect smallerRelation,
...
+        var subQueries = new HashMap<LogicalPlan, SelectSymbol>();
+        subQueries.put(smallerSideIdLookup, lookUpQuery);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/optimizer/rule/EquiJoinToLookupJoin.java:184]:
```suggestion
        var subQueries = Map.of(smallerSideIdLookup, lookUpQuery);
```

File: server/src/test/java/io/crate/integrationtests/LookupJoinIntegrationTest.java [hunk 10]
```diff
@@ -22,45 +22,68 @@
...
 
-    @Before
-    public void setup() throws Exception {
+    @UseRandomizedOptimizerRules(0)
+    @UseHashJoins(1)
+    @Test
+    public void test_lookup_join_with_two_collect_operators() throws Exception {
         execute("create table doc.t1 (id int) with(number_of_replicas=0)");
-        execute("insert into doc.t1 (id) select b from generate_series(1,10000) a(b)");
+        execute("insert into doc.t1 (id) select b from generate_series(1,20000) a(b)");
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/integrationtests/LookupJoinIntegrationTest.java:40]: Why the change here? Shouldn't the ratio remain the same? The test case is getting a bit expensive

Comment (mkleen) [reply server/src/test/java/io/crate/integrationtests/LookupJoinIntegrationTest.java:40]: That was related to the filter in the query. It's fixed now. I adapted the query.

File: server/src/main/java/io/crate/planner/Planner.java [hunk 11]
```diff
@@ -235,6 +237,10 @@ public Plan plan(AnalyzedStatement analyzedStatement, PlannerContext plannerCont
         return analyzedStatement.accept(this, plannerContext);
     }
 
+    public BiFunction<LogicalPlan, PlannerContext, LogicalPlan> optimize() {
```
Comment (mkleen) [inline server/src/main/java/io/crate/planner/Planner.java:240]: I don't like exposing the optimize function like this, but I don't have a better idea. Any suggestions ?

File: server/src/main/java/io/crate/action/sql/Session.java [hunk 12]
```diff
@@ -220,7 +220,8 @@ public void quickExec(String statement, ResultReceiver<?> resultReceiver, Row pa
             params,
             cursors,
             currentTransactionState,
-            new PlanStats(nodeCtx, txnCtx, tableStats)
+            new PlanStats(nodeCtx, txnCtx, tableStats),
+            planner::optimize
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/action/sql/Session.java:224]: I think as a follow up we could look into having a `planner.createContext(...)`, to keep the `optimize` internal and make the `PlannerContext` constructor package private.

But not sure of the implications, and if it will really improve anything but I think it might be worth a shot.

**Review by mfussenegger — APPROVED** [review]
I think this deserves another release notes entry, otherwise lgtm

**Comment by mkleen** [comment]
retest this please
